### ISSUE TITLE
shUnit2

### ISSUE THREAD

**Issue by golharam** [SimonKagstrom/kcov#133]
I have a shell script which is an analysis pipeline.  I've written unit tests that call the shell script using shUnit2.  I'm now trying to use kcov on the unit tests to see how much of the unit tests cover the pipeline.  However, I'm only getting coverage information for the unit tests and not the actual pipeline script.  How do I get coverage information for the actual script I want to test?

Edit: I've added kcov within my unit test scripts when calling the actual script I want to test.  When I do it this way, data is not collected between each call to my script, only the last run.  So it still seems that there is an incompatibility between shUnit2 and kcov.

**Comment by SimonKagstrom** [comment]
Which version of kcov are you using? There shouldn't be any incompatibility between kcov and shunit that I know of.

Anyway, one thing you can try is the --bash-handle-sh-invocation switch, although from your description it doesn't appear to be quite that problem you're facing.

**Comment by golharam** [comment]
I'm using 3.0.  I can reproduce my problem in a simple example.

The script I want to unit test is testkCov.sh:
```
#!/bin/bash

while [[ $# > 0 ]]
do
        key="$1"
        case $key in
                -s|--sample)
                SAMPLE="$2"
                shift # past argument
                ;;

                *)
                echo "Unknown option: $key"
                exit -1
                ;;
        esac
        shift   # past argument or value
done

# Check for required parameters here.
# If required parameters are not found, display help
if [ -z "${SAMPLE// }" ]
then
        echo "SAMPLE not set"
else
        echo "SAMPLE set"
fi
```
My unit test script is kcov_unitTests.sh:
```
#!/bin/sh
# vim:et:ft=sh:sts=2:sw=2

# This is called once before any tests are run
oneTimeSetUp()
{
  rm -rf coverage 2>/dev/null
  mkdir coverage
  kcov=$(readlink -f "kcov-30/bin/kcov")
  coverage_dir=$(readlink -f "coverage")
  kcovCmd="${kcov} ${coverage_dir}"

  script="${kcovCmd} $(readlink -f "./testkCov.sh")"
  unset kcov coverage_dir kcovCmd
}

#-----------------------------------------------------------------------------
# suite tests
# All tests should check:
#  return value
#  stdout, stderr
#  and any files created

testRunWithNoParameters()
{
  ${script} >/dev/null
  rtrn=$?

  assertTrue "Expected returned value of 0, but got ${rtrn}." "[ ${rtrn} -eq 0 ]"
}

XtestInvalidOption()
{
  ${script} -a >/dev/null
  rtrn=$?

  assertTrue "Expected 255, but got ${rtrn}." "[ ${rtrn} -eq 255 ]"
}

XtestSample()
{
  ${script} -s sample >/dev/null
  rtrn=$?

  assertTrue "Expected 0, but got ${rtrn}." "[ ${rtrn} -eq 0 ]"
}

# load and run shUnit2
[ -n "${ZSH_VERSION:-}" ] && SHUNIT_PARENT=$0
. shunit2/source/2.1/src/shunit2

```

If I only run the first unit test, I see coverage for the line 'echo "SAMPLE not set"'.   If I run both the first and last test (testSample) by removing the X, I no longer see coverage for the line above, instead I only see coverage for the second test.

Oddly enough in a different test script I don't see this behaviour, so I'm not sure what is causing this.

**Comment by golharam** [comment]
After playing around, I found the difference.   In my unit test script:

```
#!/bin/sh
# vim:et:ft=sh:sts=2:sw=2

# This is called once before any tests are run
oneTimeSetUp()
{
  echo "***oneTimeSetUp()***"
  echo "- Removing coverage dir"
  echo "- Creating coverage dir"

  rm -rf coverage
  mkdir coverage

  echo "- Setting up coverage collection"
  kcov=$(readlink -f "kcov-30/bin/kcov")
  coverage_dir=$(readlink -f "coverage")
  kcovCmd="${kcov} ${coverage_dir}"

  echo "- Setting up script to test"
  script="${kcovCmd} $(readlink -f "../QCPipeline/rnaseqqc_pipeline.new.sh")"
  echo "- $script"
  unset kcov coverage_dir kcovCmd

  echo "- Running tests"
  echo
}

#-----------------------------------------------------------------------------
# suite tests
# All tests should check:
#  return value
#  stdout, stderr
#  and any files created

testZero()
{
  #kcov-30/bin/kcov coverage ../QCPipeline/rnaseqqc_pipeline.new.sh 0
  ${script} 0
  rtrn=$?

  assertTrue "Expected returned value of 0, but got ${rtrn}." "[ ${rtrn} -eq 0 ]"
}

testOne()
{
  #kcov-30/bin/kcov coverage ../QCPipeline/rnaseqqc_pipeline.new.sh 1
  ${script} 1
  rtrn=$?

  assertTrue "Expected returned value of 0, but got ${rtrn}." "[ ${rtrn} -eq 0 ]"
}

# load and run shUnit2
[ -n "${ZSH_VERSION:-}" ] && SHUNIT_PARENT=$0
. shunit2/source/2.1/src/shunit2

```

If I use a fullpath to kcov, the coverage data between runs doesn't work, however if I use a relative path, all is good.  That's the only change I need to make to my test script to reproduce the issue I'm seeing.

**Comment by SimonKagstrom** [comment]
Thanks for the detailed reports! I don't have time today, but I'll try to reproduce the problem as soon as possible. If it's the full-path problem, I think it should be a fairly easy fix.

**Comment by golharam** [comment]
bumping...

**Comment by SimonKagstrom** [comment]
Sorry about the delay. I tried without being able to reproduce it, but it turns out this is actually a duplicate of Issue #130, which was fixed with eec0c67 in master. Without that commit, I get your problem, with it I don't see it. So if you checkout the git HEAD, it should work again.

I suppose I should release v31 sometime soon!
